### PR TITLE
refactor(ci): provide powerful github piplines

### PR DESCRIPTION
**Pull request by Xunzhuo**
Signed-off-by: Xunzhuo <mixdeers@gmail.com>

refactor(ci): provide powerful github piplines and scripts

``` 
Layotto commands 👀: 

A fast and efficient cloud native application runtime 🚀.
Commands below are used in Development 💻 and GitHub workflow 🌊.

Usage: make <TARGETS> <OPTIONS> ...

Targets:
  build               Build layotto for host platform.
  build.multiarch     Build layotto for multiple platforms. See option PLATFORMS.
  image               Build docker images for host arch.
  image.multiarch     Build docker images for multiple platforms. See option PLATFORMS.
  push                Push docker images to registry.
  push.multiarch      Push docker images for multiple platforms to registry.
  app                 Build app docker images for host arch. [`/docker/app` contains apps dockerfiles]
  app.multiarch       Build app docker images for multiple platforms. See option PLATFORMS.
  wasm                Build layotto wasm for linux arm64 platform.
  wasm.multiarch      Build layotto wasm for multiple platform.
  wasm.image          Build layotto wasm image for multiple platform.
  wasm.image.push     Push layotto wasm image for multiple platform.
  check               Run all go checks of code sources.
  check.lint          Run go syntax and styling of go sources.
  check.unit          Run go unit test.
  check.style         Run go style test.
  style.coverage      Run coverage analysis.
  style.deadlink      Run deadlink check test.
  style.quickstart    Run quickstart check test.
  integrate.wasm      Run integration test with wasm.
  integrate.runtime   Run integration test with runtime.
  format              Format layotto go codes style with gofmt and goimports.
  clean               Remove all files that are created by building.
  all                 Run format codes, check codes, build Layotto codes for host platform with one command
  help                Show this help info.

Options:
  BINS         The binaries to build. Default is all of cmd.
               This option is available when using: make build/build.multiarch
               Example: make build BINS="layotto_multiple_api layotto"
  IMAGES       Backend images to make. Default is all of cmds.
               This option is available when using: make image/image.multiarch/push/push.multiarch
               Example: make image.multiarch IMAGES="layotto_multiple_api layotto"
  PLATFORMS    The multiple platforms to build. Default is linux_amd64 and linux_arm64.
               This option is available when using: make build.multiarch/image.multiarch/push.multiarch
               Example: make image.multiarch IMAGES="layotto" PLATFORMS="linux_amd64 linux_arm64"
               Supported Platforms: linux_amd64 linux_arm64 darwin_amd64 darwin_arm64
```

provide tree flow pipelines in layotto like below:

Release Pipeline Demo: https://github.com/Xunzhuo/layotto/actions/runs/2213418042

PR Pipeline Demo: https://github.com/Xunzhuo/layotto/actions/runs/2213411502

Dev Pipeline Demo: https://github.com/Xunzhuo/layotto/actions/runs/2213433930

image

we can control this by workflow conditions, it is better to not set it as a required job. (Job Name can be changed and some changes should not always need to run the required workflow, like changes in docs or github files)

### CODE REVIEW

**Comment by seeflood** [comment]
The emoji is so cooooool 
image

**Comment by seeflood** [comment]
wtf !! 这就是传说中的 devops 吗!
image

File: .github/workflows/pr-title-checker.yml [hunk 1]
```diff
@@ -1,6 +1,7 @@
-name: "PR Title Checker"
+name: PR Title Checker
+
 on:
-  pull_request_target:
+  pull_requestt:
```
Comment (seeflood) [inline .github/workflows/pr-title-checker.yml:4]: find a typo here  :)

**Review by seeflood — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-title-checker.yml (hunk 1 above)
Comment (Xunzhuo) [reply .github/workflows/pr-title-checker.yml:4]: Got it thanks.

File: spec/proto/runtime/v1/appcallback.pb.go [hunk 2]
```diff
@@ -8,14 +8,15 @@ package runtime
 
 import (
 	context "context"
+	reflect "reflect"
+	sync "sync"
+
```
Comment (seeflood) [inline spec/proto/runtime/v1/appcallback.pb.go:13]: A little problem here: these `.pb.go` files are auto-generated by tools and will break the linter rules everytime it's re-generated. 
So, can we configurate the linter to ignore them?

File: spec/proto/runtime/v1/runtime.pb.go [hunk 3]
```diff
@@ -8,15 +8,16 @@ package runtime
 
 import (
 	context "context"
+	reflect "reflect"
+	sync "sync"
+
```
Comment (seeflood) [inline spec/proto/runtime/v1/runtime.pb.go:13]: Same as above

**Review by seeflood — CHANGES_REQUESTED** [review]
It seems that the deadlink checker didn't run :
image

image

**Review by seeflood — CHANGES_REQUESTED** [review]
Neither do quickstart checker:
image

**Comment by seeflood** [comment]
@ZLBer @stulzq @wenxuwan @MoonShining @zhenjunMa Could u help review this PR?
This PR refactor the workflow and add advanced scripts to make local development easier

**Comment by seeflood** [comment]
After this pr merged, we can add a "portability test" pipeline which try to use same client code to invoke different components

**Comment by Xunzhuo** [comment]
> After this pr merged, we can add a "portability test" pipeline which trys to use same client code to invoke different components

LGTM, we can reduce duplicated works if we can test different components with the same client.

File: pkg/grpc/dapr/proto/common/v1/common.pb.go [hunk 4]
```diff
@@ -12,12 +12,13 @@
 package common
 
 import (
+	reflect "reflect"
+	sync "sync"
+
```
Comment (seeflood) [inline pkg/grpc/dapr/proto/common/v1/common.pb.go:17]: The `.pb.go` files under `pkg/grpc/dapr/proto/common/v1/` directory are automatically generated too.
Can we ignore all `.pb.go` files? If not, we can ignore the directory `pkg/grpc/dapr/proto/common/v1/`

**Review by seeflood — CHANGES_REQUESTED** [review]

**Review by seeflood — CHANGES_REQUESTED** [review]
Could u also modify the doc [Explanation for github workflow](https://mosn.io/layotto/#/en/development/explanation-for-github-workflow) ?
Including :
- Explain each pipeline
- What should we do to pass the linter? For example, how to format the code?

I'm worried that contributors don't know how to fix linter issues without a doc.

**Comment by Xunzhuo** [comment]
> It seems that we have duplicated title validation?
> 
> image
> 
> 

no the duplicated workflow u see, is the merged pr title workflow, once the new workflow get merged, the older workflow will disappear

**Comment by Xunzhuo** [comment]
> Could u also modify the doc [Explanation for github workflow](https://mosn.io/layotto/#/en/development/explanation-for-github-workflow) ?
> 
> Including :
> 
> - Explain each pipeline
> 
> - What should we do to pass the linter? For example, how to format the code?
> 
> 
> 
> I'm worried that contributors don't know how to fix linter issues without a doc.

Good point,I want to write docs to explain these too.

**Comment by seeflood** [comment]
> > It seems that we have duplicated title validation?
> > image
> 
> no the duplicated workflow u see, is the merged existed pr title workflow, once the new workflow get merged, the older workflow will disappear

ok

File: pkg/grpc/dapr/proto/common/v1/common.pb.go (hunk 4 above)
Comment (Xunzhuo) [reply pkg/grpc/dapr/proto/common/v1/common.pb.go:17]: All .pb.go files have been ignored.

File: docs/zh/development/release-guide.md [hunk 5]
```diff
@@ -1,29 +1,55 @@
...
+#### Step5: 确认发布
 
-- [ ] 点按钮，发布
-- [ ] 群里发个消息通知下大家  
-- [ ] 发布完成后，创建[下次迭代的roadmap](https://github.com/mosn/layotto/projects) （可能已经有了），可以把上个版本没做完的事项挪进来
-- [ ] 如果有相应的 SDK 需要发布，参考上述流程，在 SDK 仓库做发布，并上传中央仓库 (比如 java sdk需要上传到 Maven 中央仓库)。
\ No newline at end of file
+1. 点击发布
+2. 社区周知
+3. 创建[下次迭代的 Roadmap](https://github.com/mosn/layotto/projects) ，迁移上个版本未完成的任务
```
Comment (seeflood) [inline docs/zh/development/release-guide.md:54]: 这句需要修改下，现在每次发布之后不需要关闭 Roadmap 了，新版本可以复用之前的 Roadmap
比如改成:

3. 检查 [Roadmap](https://github.com/mosn/layotto/projects) ，修改上个版本未完成的任务，把 milestone 改成下个版本

**Review by seeflood — CHANGES_REQUESTED** [review]

**Review by seeflood — APPROVED** [review]
lgtm

**Comment by seeflood** [comment]
image
Some quickstart may fail due to the changes in Makefile. We can submit new PR to fix them after merging this

**Review by wenxuwan — APPROVED** [review]

**Comment by seeflood** [comment]
Thanks for this powerful new CI !

**Pull request merged**